File: guests/haskell/format.hs
{-# LANGUAGE OverloadedStrings #-}

-- The formatter door and the catalog, Haskell port — guests/rust/format.rs,
-- tools/scenes/format.steps.

import Data.Text (Text)
import Data.Time.Calendar (fromGregorian)
import Data.Time.LocalTime (TimeOfDay (..))
import KayaApp

main :: IO ()
main = kayaMain $ \app -> do
  catalog "format"
  let d = fromGregorian 2026 9 7
      t = TimeOfDay 8 30 0
  dShort <- fmtDate Short d
  dMedium <- fmtDate Medium d
  dLong <- fmtDate Long d
  tShort <- fmtTime Short t
  dt <- fmtDateTime Medium d t
  n <- fmtNumber numberOptions 1234567.891
  p <- fmtPercent numberOptions 0.256
  c <- fmtCurrency 1234567.89 "USD"
  one <- tr "items" [("count", arg (1 :: Int))]
  three <- tr "items" [("count", arg (3 :: Int))]
  hello <- tr "greeting" [("name", arg ("Ada" :: Text))]
  tag <- (.localeTag) <$> locale
  buildTx app $ do
    -- Fourteen labels and a row: taller than the default window.
    window primary [WTitle "format", WSize 540 560]
    root <-
      column
        []
        [ labelText dShort, -- label#0
          labelText dMedium, -- label#1
          labelText dLong, -- label#2
          labelText tShort, -- label#3
          labelText dt, -- label#4
          labelText n, -- label#5
          labelText p, -- label#6
          labelText c, -- label#7
          labelText one, -- label#8
          labelText three, -- label#9
          labelText hello, -- label#10
          row [] [labelText "first", spacer, labelText "last"], -- row#0, label#11, label#12
          labelText tag -- label#13
        ]
    mount root

File: guests/rust/format.rs
//! The formatter door and the catalog (tools/scenes/format.steps,
//! docs/compliance-plan.md §6): fixed inputs through every fmt call and
//! two catalog messages, asserted against what THIS platform's formatter
//! writes and against the catalog's own bytes.

use kaya::fmt::{self, Length, NumberOptions};

pub(crate) fn app(ctx: kaya::AppCtx) {
    let msgs = kaya::Messages::<()>::new();
    kaya::catalog("format");
    let d = kaya::Date::new(2026, 9, 7).unwrap();
    let t = kaya::Time::new(8, 30).unwrap();
    ctx.apply(|tx| {
        // Fourteen labels and a row: taller than the default window, which
        // GTK would otherwise let the root overflow (expect_root_fills), and
        // taller than a phone at twice the text size, so the column scrolls
        // (the iOS formatbig leg under expect_no_clipping's off-screen
        // clause, docs/flex-shrink-plan.md §4).
        tx.window(kaya::DEFAULT_WINDOW).title("format").size(540.0, 560.0);
        let root = tx
            .scroll(|tx| {
                tx.column(|tx| {
                let s = tx.signal(fmt::date(d, Length::Short));
                tx.label(s); // label#0
                let s = tx.signal(fmt::date(d, Length::Medium));
                tx.label(s); // label#1
                let s = tx.signal(fmt::date(d, Length::Long));
                tx.label(s); // label#2
                let s = tx.signal(fmt::time(t, Length::Short));
                tx.label(s); // label#3
                let s = tx.signal(fmt::date_time(d, t, Length::Medium));
                tx.label(s); // label#4
                let s = tx.signal(fmt::number(1234567.891, NumberOptions::default()));
                tx.label(s); // label#5
                let s = tx.signal(fmt::percent(0.256, NumberOptions::default()));
                tx.label(s); // label#6
                let s = tx.signal(fmt::currency(1234567.89, "USD"));
                tx.label(s); // label#7
                let s = tx.signal(kaya::tr!("items", count = 1));
                tx.label(s); // label#8
                let s = tx.signal(kaya::tr!("items", count = 3));
                tx.label(s); // label#9
                let s = tx.signal(kaya::tr!("greeting", name = "Ada"));
                tx.label(s); // label#10
                tx.row(|tx| {
                    let s = tx.signal("first");
                tx.label(s); // label#11
                    tx.spacer();
                    let s = tx.signal("last");
                tx.label(s); // label#12
                }); // row#0
                let info = fmt::locale();
                let s = tx.signal(info.tag);
                tx.label(s); // label#13
                });
            })
            .id();
        tx.mount(root);
    });

    while msgs.next(&ctx).is_some() {}
}

fn main() {
    kaya::run(app)
}
